File: programs/scala/primes.scala
import scala.collection.mutable.HashMap
import scala.collection.immutable.Vector
import scala.math.sqrt
import scala.math.pow
import scala.math.BigInt

def get_keys(n: Long, half_size:Long): Vector[Long] =
  var v = Vector.range(0L, half_size + 1L).map(i => n / (i + 1))
  val vmax = v(half_size.asInstanceOf[Int])
  v ++ Vector.range(1L, vmax).reverse

def get_sums(keys: Vector[Long]) : HashMap[Long, BigInt] =
  var hmap = HashMap.empty[Long, BigInt]
  for key <- keys do
    val k = BigInt(key)
    val value = (k * (k + 1)) / 2 - 1
    hmap.put(key, value)
  hmap 

def calculate_sums(hmap: HashMap[Long, BigInt], keys: Vector[Long], square_n: Long): HashMap[Long, BigInt] =
  var cmap = hmap.clone()
  for p <- (2L to square_n) do
    val current_sum = cmap(p - 1)
    if (cmap(p) > current_sum) {
      val p_square = p * p
      for key <- keys.takeWhile(p => p >= p_square) do
        val new_value = cmap(key) - p * (cmap(key / p) - current_sum)
        cmap.update(key, new_value)
    }
  cmap

def primes(n: Long): BigInt =
  val square_n = sqrt(n.asInstanceOf[Float]).asInstanceOf[Long]
  val half_size = square_n - 1
  val keys = get_keys(n, half_size)
  var all_sums = get_sums(keys)
  all_sums = calculate_sums(all_sums, keys, square_n)
  all_sums(n)

def add_space(i:Int, string:String) : String =
  if (i == 0){
    return string
  } else {
    return add_space(i - 1, string.concat(" "))
  }

def format(dist:Int, string: String): String =
  add_space(dist - string.length(), string)

@main def main() =
  println("Power | Time (µs) | Resultat")
  println("========================================")
  for power <- (1 to 8) do
    val start = System.nanoTime
    val n = pow(10.0, power.asInstanceOf[Double]).asInstanceOf[Long]
    val result = primes(n)
    val duration = ((System.nanoTime - start) / 1e3d).asInstanceOf[Int]
    println(s"${format(5, s"$power")} | ${format(9, s"$duration")} | $result")

File: programs/java/primes.java
import java.util.HashMap;
import java.util.Vector;
import java.lang.Math;

public class Main{

    static Vector<Integer> get_keys(Integer n, Integer half_size){
        Vector<Integer> V = new Vector<>();
        for (Integer i = 0; i < half_size; i++){
            V.add(n / (i + 1));
        }
        Integer vmax = V.get(V.size() - 1) - 1;
        for (Integer i = vmax; i > 0; i--){
            V.add(i);
        }
        return V;
    }

    static HashMap<Integer, Long> get_sums(Vector<Integer> keys){
        HashMap<Integer, Long> hmap = new HashMap<>();
        for (int key: keys){
            long big_key = key;
            long value = (big_key * (big_key + 1)) / 2 - 1;
            hmap.put(key, value);
        }
        return hmap;
    }

    static HashMap<Integer, Long> calculate_sums(HashMap<Integer, Long> hmap, Vector<Integer> keys, int square_n){
        for (Integer p = 2; p < square_n + 1; p++){
            long current_sum = hmap.get(p - 1);
            if (hmap.get(p) > current_sum){
                long p_square = p * p;
                for (int key: keys){
                    if (key < p_square){
                        break;
                    }
                    long new_value = hmap.get(key) - (long)p * (hmap.get(key / p) - current_sum);
                    hmap.put(key, new_value);
                }
            }
        }
        return hmap;
    }

    static long primes(int n){
        int square_n = (int) Math.sqrt(n);
        assert square_n * square_n <= n && (square_n + 1) * (square_n + 1) > n;
        int half_size = square_n - 1;
        Vector<Integer> keys = Main.get_keys(n, half_size);
        HashMap<Integer, Long> all_sums = Main.get_sums(keys);
        all_sums = Main.calculate_sums(all_sums, keys, square_n);
        return all_sums.get(n);
    }

    static String format(int dist, String value){
        return value + " ".repeat(dist - value.length());
    }

    public static void main(String[] args){
        Main App = new Main();
        System.out.println("Power | Time (µs) | Resultat");
        System.out.println("========================================");
        primes(100);
        for (int i = 1; i < 9; i++){
            long start = System.nanoTime();
            long result = primes((int) Math.pow(10, i));
            long duration = (System.nanoTime() - start) / 1_000;
            System.out.println(
                Main.format(5, String.valueOf(i)) + " | " +
                Main.format(9, String.valueOf(duration)) + " | " +
                String.valueOf(result)
            );
        }
    }
}
